Fixed #31826 -- Made AlterField operation a noop when adding db_column.

iurisilvio opened the pull request:
https://code.djangoproject.com/ticket/31826

felixxm commented:
@iurisilvio Thanks :+1: I was able to reproduce this behavior only on SQLite. IMO we shouldn't add this optimization to the `AlterField` operations. It looks that both issues (ticket-31826, ticket-31825) can be fixed in the SQLite backend, see [comment](https://github.com/django/django/pull/13236#issuecomment-665584461).

iurisilvio commented:
Same thing about sqlite3 on this PR, it fails with foreign keys on postgres.

I moved my implementation to base schema `alter_field`. I don't have to copy it to sqlite3 schema because their behavior is different only on field name changes.

Also, added tests to the foreign key case.

This PR is ready for another review, but it conflicts with #13236, so I prefer the RenameField PR merged first.

iurisilvio commented:
I updated this PR with current master and moved the change to the `_field_should_be_altered`.

felixxm commented on django/db/backends/base/schema.py:
```diff
@@ -1037,11 +1037,28 @@ def _field_indexes_sql(self, model, field):
+            if changed_kwargs == {'db_column'}:
+                return False
+
+        return True
```
We've already compare columns in:
```python
old_field.column != new_field.column or
```
so we can simply ignore `db_column` in kwargs, e.g.
```python
def _field_should_be_altered(self, old_field, new_field):
    _, old_path, old_args, old_kwargs = old_field.deconstruct()
    _, new_path, new_args, new_kwargs = new_field.deconstruct()
    # Ignore db_column to not alter when changing only a db_column but it's
    # the same as a field name.
    old_kwargs.pop('db_column', None)
    new_kwargs.pop('db_column', None)
    # Don't alter when changing only a field name.
    return (
        old_field.column != new_field.column or
        (old_path, old_args, old_kwargs) != (new_path, new_args, new_kwargs)
    )
```

iurisilvio replied:
Great solution! Thanks! I didn't rethink the past implementation. Updated. :+1:

felixxm commented:
@iurisilvio Thanks :+1: I pushed small edits.

felixxm commented:
buildbot, test on oracle.
